core/src/main/java/software/amazon/awssdk/core/util/Mimetypes.java

```diff
@@ -0,0 +1,174 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.util;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringTokenizer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.utils.IoUtils;
+
+/**
+ * Utility class that maintains a listing of known Mimetypes, and determines the
+ * mimetype of files based on file extensions.
+ * <p>
+ * This class is obtained with the {#link {@link #getInstance()} method that
+ * recognizes loaded mime types from the file <code>mime.types</code> if this
+ * file is available at the root of the classpath. The mime.types file format,
+ * and most of the content, is taken from the Apache HTTP server's mime.types
+ * file.
+ * <p>
+ * The format for mime type setting documents is:
+ * <code>mimetype + extension (+ extension)*</code>. Any
+ * blank lines in the file are ignored, as are lines starting with
+ * <code>#</code> which are considered comments.
+ *
+ * @see <a href="https://github.com/apache/httpd/blob/trunk/docs/conf/mime.types">mime.types</a>
+ */
+public final class Mimetypes {
+
+    /** The default XML mimetype: application/xml */
+    public static final String MIMETYPE_XML = "application/xml";
+
+    /** The default HTML mimetype: text/html */
+    public static final String MIMETYPE_HTML = "text/html";
+
+    /** The default binary mimetype: application/octet-stream */
+    public static final String MIMETYPE_OCTET_STREAM = "application/octet-stream";
+
+    /** The default gzip mimetype: application/x-gzip */
+    public static final String MIMETYPE_GZIP = "application/x-gzip";
+
+    public static final String MIMETYPE_TEXT_PLAIN = "text/plain; charset=UTF-8";
+
+    private static final Logger LOG = LoggerFactory.getLogger(Mimetypes.class);
+
+    private static final String MIME_TYPE_PATH = "software/amazon/awssdk/core/util/mime.types";
+
+    private static final ClassLoader CLASS_LOADER = ClassLoaderHelper.classLoader();
+
+    private static volatile Mimetypes mimetypes;
+
+    /**
+     * Map that stores file extensions as keys, and the corresponding mimetype as values.
+     */
+    private Map<String, String> extensionToMimetypeMap = new HashMap<>();
+
+    private Mimetypes() {
+    }
+
+    /**
+     * Loads MIME type info from the file 'mime.types' in the classpath, if it's available.
+     */
+    public static Mimetypes getInstance() {
+
+        if (mimetypes == null) {
+            synchronized (Mimetypes.class) {
+                if (mimetypes == null) {
+                    mimetypes = new Mimetypes();
+                    Optional.ofNullable(CLASS_LOADER).map(loader -> loader.getResourceAsStream(MIME_TYPE_PATH)).ifPresent(
```

Comment:
Why isn't this done in the constructor? If it were the `extensionToMimetypeMap` would be effectively immutable.